pkg/action/hooks.go

```diff
@@ -107,6 +124,17 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 	return nil
 }
 
+// getHookLog gets the log from the pod associated with the given hook, which is expected to be a test hook.
+func getHookLog(client kubernetes.Interface, rel *release.Release, hook *release.Hook) (release.HookLog, error) {
+	req := client.CoreV1().Pods(rel.Namespace).GetLogs(hook.Name, &v1.PodLogOptions{})
+	responseBody, err := req.DoRaw(context.Background())
+	if err != nil {
+		var nothing release.HookLog
+		return nothing, errors.Wrapf(err, "unable to get pod logs for %s", hook.Name)
```

Comment:
Instead of returning an empty structure... Why not return the response body? That way we can check for `nil, err` rather than `release.HookLog{}, err`.

An empty struct generally means the object can still be handled. Returning a `nil` object is usually better because it provides context to the caller that the should not perform any business logic with this object.

Reply:
> Instead of returning an empty structure... Why not return the response body?

I'm not sure I'm following here, but if we've promised (in the name and type signature of `getHookLog`) to return a `HookLog`, it wouldn't feel honest to return something that's definitely not a hook log.

Note that we're not returning an empty structure, but rather the empty string.

> That way we can check for `nil, err` rather than `release.HookLog{}, err`
>
> An empty struct generally means the object can still be handled. Returning a `nil` object is usually better because it provides context to the caller that the should not perform any further business logic with this object.

I must admit that I may not fully understand the point and its implications here, but I'd expect `getHookLog` to be idiomatically called like this:

```go
hookLog, err := getHookLog(client, rl, h)
if err != nil {
	// Handle error …
}
```

That is, if the second element of `getHookLog`'s return value (`err`) is not `nil`, it shouldn't matter what the first element (`hookLog`) is. Furthermore, the first element _can't_ be `nil`, because `nil` is not a member of the `HookLog` (`string`) type.

Finally, I'd like to mention that I find Go's standard error handling idiom – which uses a product type (_T_ × `error`) to represent what is, for all intents and purposes, actually a sum type (_T_ + `error`) – pretty frustrating to work with. I don't know of any "perfect" way to use it; the `nothing` idiom is basically the least bad one AFAIK.

Reply:
WDYT, @bacongobbler?

Reply:
The pattern of returning `nil, err` when an error occurs is the standard idiom across Go's standard library. The ONLY exception is when the user can actually do something useful with the data, like when the developer calls `ReadLine` and it hits EOF. In that case it should return `line, io.EOF` because the user can still so something with the rest of the line and catch `io.EOF`.

In this case there isn't anything the user can do with the hook log, so we should return a `nil` object.

Hope this helps.

Reply:
But we literally cannot return `nil`, because the return type is `(string, error)` and `nil` is not a member of `string`.

Surely the standard idiom in general is not returning `nil, err`, but rather returning `<the zero value>, err` (e.g. `"", err`, `0, err` etc)? Some examples of functions in the standard library that do so are [`os.MkdirTemp`](https://pkg.go.dev/os@go1.17.3#MkdirTemp), [`os.ReadLink`](https://pkg.go.dev/os@go1.17.3#Readlink) and [`bytes.Reader.Read`](https://pkg.go.dev/bytes@go1.17.3#Reader.Read). As far as I can tell, `getHookLog` is idiomatically compliant in that aspect.

Reply:
I don't understand why you're returning an untyped raw response body in the first place. This might have more to do with `req.DoRaw` than the actual implementation details but in general passing raw response bodies back to the user is asking for trouble.

Most functions provide some type of higher-level abstraction so that the caller can actually do something useful with the response body...

Reply:
I have now pushed a commit that reverts from `req.DoRaw` to `req.Stream`.